### PR TITLE
Improve logging/debugging when select() fails

### PR DESCRIPTION
**Pull request by ijc25**
Retry when appropriate and log the failure reason for other failures.

### CODE REVIEW

File: src/pci_virtio_sock.c [hunk 1]
```diff
@@ -340,6 +340,34 @@ static int max_fd(int a, int b)
...
+	if (rc >= 0) return rc;
+
+	/*
+	 * http://pubs.opengroup.org/onlinepubs/009695399/functions/select.html
+	 * lists EINTR, EBADF and EINVAL. EINTR is recoverable and should be
+	 * retried.
+	 */
+	if (errno == EINTR) return -1;
+	/*
+	 * OS select(2) man page lists EAGAIN in addition to the above.
```
Comment (djs55) [inline src/pci_virtio_sock.c]: s/OS/OSX/ ?

Comment (ijc25) [inline src/pci_virtio_sock.c]: Yep, will fix, thanks!

**Comment by djs55** [comment]
LGTM

Out of interest, do you happen to know what `FD_SETSIZE` is on OSX? Is it worth compiling with `-D_DARWIN_UNLIMITED_SELECT`? The OSX manpage has:

>     select() now returns with errno set to EINVAL when nfds is greater than FD_SETSIZE.  Use a smaller value for nfds
>     or compile with -D_DARWIN_UNLIMITED_SELECT.

**Comment by ijc25** [comment]
> LGTM

Thanks!

> Out of ....

`grep -r FD_SETSIZE /usr/include/` throws up a lot of noise but it looks like it is in the vicinity of 1024.

**Comment by justincormack** [comment]
LGTM

**Comment by justincormack** [comment]
I would just use kqueue if you want bigger sets than select allows, it will perform better anyway

**Comment by ijc25** [comment]
> Is it worth compiling with -D_DARWIN_UNLIMITED_SELECT?

To answer @djs55's actual question: Reading `/usr/include/sys/_select.h` implies the only reason it isn't the default is for "UNIX conformance" (which seems like an odd choice, but oh well), but I'm not sure what the other unspoken trade-offs there might be.

I think I would want to make that change on the master branch, since the use of `select` is not specific to vsock. I'll raise another PR in a moment with that change.

**Pull request merged**